Pull request test

Hawk94 opened the pull request:
Testing out pull requests. example.txt highlights basic usage of the zoo. There are comments on class functions detailing how I might extend the project. 

Have I used 'super' correctly on the BaseSpecies class extensions??

seddonym commented on README.rst:
```diff
@@ -19,8 +19,34 @@ Zoo Game
 Bonus task
 ----------
 2.  Each cage needs a reference number so the zoo keeper doesn't get confused.  Automatically generate a reference number when building each cage.  The first cage should be 1, the second 2, etc.
+
```
I think you still need to remove the .DS_Store file from version control.

seddonym commented on zoo/example.txt:
```diff
@@ -0,0 +1,24 @@
+
+Set up of a simple zoo with two cages for testing.
+
+from zoo.zoo import Zoo, Cage, BaseSpecies, Lion, Hyena, Wildebeest, Gazelle
```
`zoo.zoo` seems verbose.  See if you can work out how to make it work like this:

    from zoo import Zoo, etc.

seddonym commented on zoo/example.txt:
```diff
@@ -0,0 +1,24 @@
+
+from zoo.zoo import Zoo, Cage, BaseSpecies, Lion, Hyena, Wildebeest, Gazelle
+zoo = Zoo('Test Zoo')
+zoo.__repr__  # Displays the name of the zoo alongside the current number of cages and animals
```
Just typing `zoo` should do it.  But if you want to get the repr within Python (rare), you should do `repr(zoo)`.

seddonym commented on zoo/example.txt:
```diff
@@ -0,0 +1,24 @@
+from zoo.zoo import Zoo, Cage, BaseSpecies, Lion, Hyena, Wildebeest, Gazelle
+zoo = Zoo('Test Zoo')
+zoo.__repr__  # Displays the name of the zoo alongside the current number of cages and animals
+one_cage = zoo.add_cage('One Cage')  # Cages can be created directly from a Zoo object passing only their name.
```
Nice, but I reckon it's a clearer API to instantiate cages rather than add cages as strings.  Example:

    one_cage = Cage('One Cage')
    zoo.add_cage(one_cage)

Another way to do it could be for the cages to know about the zoo, rather than vice versa.  That would look more like:

    zoo = Zoo('My zoo')
    cage = Cage('My cage', zoo=zoo)

Have a think about the strengths and weaknesses of each approach.

seddonym commented on zoo/example.txt:
```diff
@@ -0,0 +1,24 @@
+another_cage = zoo.add_cage('Another Cage')
+
+one_cage.__repr__  # Displays the cage name alongside the animals that are currently in it.
+one_cage.add_animal(Lion('Lilly'))  # Animals can be created directly from a cage or independently.
```
Nice.  Again, another way could be:

     lion = Lion('Lilly')
     lion.add_to_cage(one_cage)

Have a think about strengths and weaknesses of either.

seddonym commented on zoo/example.txt:
```diff
@@ -0,0 +1,24 @@
+one_cage.__repr__  # Displays the cage name alongside the animals that are currently in it.
+one_cage.add_animal(Lion('Lilly'))  # Animals can be created directly from a cage or independently.
+one_cage.add_animal(Hyena('Henny'))  # In this case the cage_relationships method checks whether animals are eaten. In this case there are none.
+one_cage.add_animal(Wildebeest('Willy'))  # In this case Willy is instantly eaten.
```
You could simplify your API to begin with by just doing something like:

    willy = Wildebeest('Willy')
    one_cage.add_animal(willy)

Then you don't need the get_animal method on the zoo class.

seddonym commented on zoo/example.txt:
```diff
@@ -0,0 +1,24 @@
+one_cage.add_animal(Wildebeest('Willy'))  # In this case Willy is instantly eaten.
+willy = zoo.get_animal('Willy')  # Zoo's have a get_animal helper method to make finding animals easy.
+willy.get_status()  # The get_status method returns a human readable status of an animal object. Willy is unfortunately 'Dead'
+willy.sit_rep  # Sit_rep records who an animal was eaten by.
```
`sit_rep` is a weird naming, can you think of an improved API for this?

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+"""You are a zoo keeper. Write a set of objects that simulates a simple zoo."""
+
+ANIMAL_STATUS = {
```
This is nice.  Even nicer:

    class Animal:
         STATUS_ALIVE = 'AL'
         STATUS_DEAD = 'DE'
         STATUSES = {
             STATUS_ALIVE: 'Alive',
             STATUS_DEAD: 'Dead',
         }

Namespacing it in the class allows you to refer to the statuses like:

    Animal.STATUS_ALIVE

seddonym replied:
Ah I've seen you've done this elsewhere :)

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+}
+
+class Zoo(object):
+    """Creates a Zoo object with a name and an empty list for cages."""
```
The docstring here is more about the __init__ method than the Zoo itself.  Potentially you could do something like:

    class Zoo:
        """A Zoo is a collection of cages with animals in them.
        """

Note that you don't need `(object)` in python 3.

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+class Zoo(object):
+    """Creates a Zoo object with a name and an empty list for cages."""
+
+    zoo_cages = 0
```
This is defined on the class scope, which is not correct.  You should instead define it on the instance (probably when you instantiate it).

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+        return self.name
+
+    def __repr__(self):
+        """Class representation, returned when class object is called."""
```
Actually that isn't what repr does.  http://stackoverflow.com/questions/7784148/understanding-repr-function-in-python

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+        self.cages.append(cage)
+        self.zoo_cages += 1
+        return cage
+        
```
Extra line - are you running flake8?

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+        """Cage objects are added to a Zoo's cages list."""
+        cage = Cage(cage_name)
+        self.cages.append(cage)
+        self.zoo_cages += 1
```
Can you think of another way to get the number of cages in the zoo, without needing to store it manually?

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+
+    def remove_cage(self, cage_name):
+        """The Zoo should keep track of how many cages are in existence."""
+        cage = self.get_cage(cage_name)
```
Think about rewriting this to accept a cage object rather than a string.

Also, see if you can work out a more elegant way to remove the cage from the zoo.

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+        move_to = self.cage(cage_to)
+        moved_animal = self.animal(animal_name)
+
+        move_to.cage_contents.append(moved_animal)
```
One for discussion, but there is an interesting issue with the data model that this illustrates.  Can you work out what it is?

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+                    return i
+
+    def move_animal(self, cage_from, cage_to, animal_name):
+        """An animal can be moved between cages, but only within a Zoo."""
```
The docstring should make it clear what data types this method expects.  Are the cage_from and cage_to arguments Cage objects or strings?

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+        self.cage_contents = []
+        if zoo:
+            self.zoo = zoo
+        Cage.instances += 1
```
This is a bit idiosyncratic.  Let's have a conversation about whether it's necessary, and if there are other ways to achieve it.

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+    """Create a Lion."""
+
+    def __init__(self, name):
+        super(self.__class__, self).__init__(name)
```
In Python 3, you can just do `super().__init__(name)`.  I'd change this everywhere in the file.

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+                return None
+
+
+class BaseSpecies(object):
```
I would say that a better name for this class would be `Animal`.  Species is a characteristic of an animal, but the animal is the class of object.  We can discuss this more.

seddonym commented on zoo/zoo.py:
```diff
@@ -0,0 +1,281 @@
+        raise NotImplementedError
+
+
+class Lion(BaseSpecies):
```
See if you can think of a way of configuring these subclasses without extending any methods (just setting attributes on the class).

Hawk94 commented:
@seddonym  Hey David. I have written a zoo systems test that addresses the points raised in the task outline. I have also run the tests locally and run the coverage report. Take a look when you get a 'mo please!
